### PR TITLE
[24068] Remove Legacy -- behavior and use of UnparsedTokens

### PR DESCRIPTION
**Pull request by Tunduk**
@baronfel It was a little bit tricky. 
Honestly, I dont't like how `System.CommandLine` works without double dash legacy behavior. It treats  double dash arguments as command arguments but it's not true. Double dash arguments should be separated from command arguments. It's impossible to create something like `dotnet vstest` with `System.CommandLine` now. 
I wanted to remove `ParseResultExtensions.GetArguments` because I don't think that we should read\modify raw values from command line. We have a whole library for that. But it can't be implemented now. 
So, I think this PR will resolve just a part of issue. It'll allow `System.CommandLine` to move forward, but it won't fix all problems in SDK 
What do you think about it?

Also fixed  some unrelated staff: 

- Fixed`dotnet test` argument arity. It should be ZeroOrOne by documentation.
- Fixed `dotnet format`  help output with subcommands. 

Thanks in advance for your review.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/sdk?pullRequest=25854) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/sdk?pullRequest=25854) before we can accept your contribution.<br/><br/>:x: Tunduk [sign now](https://cla.dotnetfoundation.org/dotnet/sdk?pullRequest=25854)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/sdk?pullRequest=25854) it.</sub>

**Review by jmarolf — COMMENTED** [review]
dotnet-format section looks good

**Comment by baronfel** [comment]
This looks great! I'll dig into test specifically tomorrow. Is there anything specific you saw that can't be solved by repeatedly gathering args via `--` and re-parsing that subset of args?

**Comment by Tunduk** [comment]
> Is there anything specific you saw that can't be solved by repeatedly gathering args via `--` and re-parsing that subset of args?

I'm not sure that I understand your question. Do you mean hack like in `src/Cli/dotnet/commands/dotnet-test/Program.cs:127` for separation command argument from `--` arguments?

File: src/Cli/dotnet/Parser.cs [hunk 1]
```diff
@@ -269,6 +268,15 @@ public override void Write(HelpContext context)
...
                 }
+                else if (command.Name.Equals(FormatCommandParser.GetCommand().Name))
+                {
+                    var argumetns = context.ParseResult.GetValueForArgument(FormatCommandParser.Arguments);
+                    new DotnetFormatForwardingApp(argumetns.Concat(helpArgs).ToArray()).Execute();
+                }
+                else if (command.Name.Equals(FsiCommandParser.GetCommand().Name))
+                {
+                    new FsiForwardingApp(helpArgs).Execute();
+                }
```
Comment (baronfel) [inline src/Cli/dotnet/Parser.cs:279]: This is great for this PR (and nothing that you;ve done incorrectly), but the more I see this the more I want to extract all of these to explicit subcommands of `System.CommandLine.Command`, so that they could override their individual help implementations/potentially do completions/etc. Making a note to myself here :)

**Comment by baronfel** [comment]
Ok I spent a good chunk of time on this yesterday and I'm really happy with it, excellent work! The problem of the test command specifically stuck with me, though, and I chatted with @jonsequitur about how we should model structures like the test command.

A refresher:

`dotnet test  [zero or one sln or project argument] [<tons of options>] -- [zero or more test application arguments] [zero or more test runner arguments]`

We can make use of custom parsing delegates to handle these three distinct arguments in a more general way, without really running afoul of `--` parsing semantics.  I've started a branch based off of this PR at [my fork](https://github.com/tunduk/sdk/compare/main...baronfel:test-parser-tweaks?expand=1) where I dug into what this looks like, and I think it has a lot of promise.  While prototyping this, however, I ran into https://github.com/dotnet/command-line-api/issues/1759. Once that's fixed I think we can actually do this!

The up-side of making custom parsers for these three arguments is
* codifying the requirements on the parsing side, which means we can get nicer parsing errors for users
* reducing the amount of custom logic during command execution
* getting more accurate help output for these kinds of arguments

**Review by baronfel — CHANGES_REQUESTED** [review]
Putting 'request changes' on this just to prevent accidental merges.

@Tunduk I am very happy with this overall, I just think we have to hold it until the linked System.CommandLine issue is resolved so that my changes to dotnet-test can be part of it as well.  I've talked to the maintainers there and they're looking into it.

Thank you so much for this - I'm very impressed by the work, and I'm very glad that you took the time to get into the details!

File: src/Cli/dotnet/Parser.cs (hunk 1 above)
Comment (Tunduk) [reply src/Cli/dotnet/Parser.cs:279]: I agree with you. 
But don't you think that it's hidden dependency? I mean a situation when DotnetFormat team add new subcommand and SDK team should add this subcommand too for correct help/completition. 
If we could deliver DotnetFormat via git submodule/nuget I think it will solve this situation.

**Comment by Tunduk** [comment]
@baronfel Sorry for long responce. Thanks a lot for your review! It was very interesting to take part in SDK project. I think I'll continue contribute in Micsrosoft projects. 

As I can see System.CommandLine issue was resolved should I do anything else with this PR?

**Comment by baronfel** [comment]
@Tunduk can you rebase your branch so that it's more up to date with `dotnet/sdk:main`? I'm ready to send a PR to your branch but because main here has advanced it looks like more changes than it really is.

**Comment by Tunduk** [comment]
@baronfel Hi!
There is still the [issue](https://github.com/dotnet/command-line-api/issues/1779) in `System.CommandLine` that blocks using custom parser feature. 
Could we postpone custom parser changes? It'll allow System.CommandLine team start work on removing legacy double dash behavior. Also, I have seen some issues that await this PR. Such as [this](https://github.com/dotnet/sdk/issues/23136)
What do you think about it?

**Comment by baronfel** [comment]
So @Tunduk at this point I think you can revert/reset to remove my commits entirely, and then we should be able to take this.
